File: py_pub_sub_pkg/py_pub_sub_pkg/publisher_hello_ros.py
# Publisher Node
#!/usr/bin/env python3
# Shebang line indicating to use Python 3 to run this script.

import rclpy
# Import the rclpy library.

from rclpy.node import Node
# Import the Node class for creating ROS 2 nodes.

from std_msgs.msg import String
# Import the String message type for publishing messages.

class PublisherNode(Node):
    # Define a class that inherits from Node, representing a ROS 2 publisher node.
    def __init__(self):
        # Constructor for the PublisherNode class.
        super().__init__('publisher_node')
        # Call the parent class constructor and name this node 'publisher_node'.

        self.publisher_ = self.create_publisher(String, 'chatter', 10)
        # Create a publisher to send String messages to the 'chatter' topic with a queue size of 10.
        
        self.timer = self.create_timer(1.0, self.publish_message)
        # Create a timer that triggers the publish_message method every 1.0 seconds.
        
        self.counter = 0
        # Initialize a counter to keep track of the number of messages published.

    def publish_message(self):
        # Method to publish messages to the 'chatter' topic.
        msg = String()
        # Create a new String message.
        
        msg.data = f"Hello ROS 2! Counter: {self.counter}"
        # Set the content of the message to include the counter value.
        
        self.publisher_.publish(msg)
        # Publish the message to the 'chatter' topic.
        
        self.get_logger().info(f"Publishing: '{msg.data}'")
        # Log the published message to the console.
        
        self.counter += 1
        # Increment the counter for the next message.

def main(args=None):
    # Define the main function.
    rclpy.init(args=args)
    # Initialize the rclpy library.
    
    node = PublisherNode()
    # Create an instance of the PublisherNode class.
    
    rclpy.spin(node)
    # Keep the node running and publishing messages periodically.
    
    node.destroy_node()
    # Destroy the node when done to clean up resources.
    
    rclpy.shutdown()
    # Shut down the rclpy library to release resources.

if __name__ == '__main__':
    main()

File: cpp_pub_sub_pkg/src/publisher_hello_ros.cpp
// Publisher Node
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class PublisherNode : public rclcpp::Node {
public:
    PublisherNode() : Node("publisher_node"), counter_(0) {
        // Create a publisher to the "chatter" topic
        publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);

        // Create a timer to publish messages every 1 second
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1), std::bind(&PublisherNode::publish_message, this));
    }

private:
    void publish_message() {
        // Create and populate the message
        auto msg = std_msgs::msg::String();
        msg.data = "Hello ROS 2! Counter: " + std::to_string(counter_);

        // Publish the message
        publisher_->publish(msg);

        // Log the message
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", msg.data.c_str());

        // Increment the counter
        counter_++;
    }

    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    int counter_;
};

int main(int argc, char **argv) {
    // Initialize the rclcpp library
    rclcpp::init(argc, argv);

    // Create a shared pointer for the publisher node
    auto node = std::make_shared<PublisherNode>();

    // Spin the node to keep it running and publishing messages
    rclcpp::spin(node);

    // Shutdown and clean up resources
    rclcpp::shutdown();
    return 0;
}